File: wc-clone/wc-go/wc_test.go
package main

import (
	"bytes"
	"io"
	"os"
	"testing"
)

func TestDefaultCount(t *testing.T) {
    file, err := os.Create("test.txt")
    if err != nil {
        t.Fatalf("Failed to create temporary file: %v", err)
    } 

    defer file.Close()
    defer os.Remove("test.txt")

    content := "Line 1\nLine 2\nLine 3\n"
    _, err = file.WriteString(content)
    if err != nil {
        t.Fatalf("Failed to write to temporary file: %v", err)
    } 

    lineCount, wordCount, byteCount, err := defaultCount("test.txt")
    if err != nil {
        t.Fatalf("Unexpected error: %v", err)
    } 

    expectedLineCount := 3
    expectedWordCount := 6
    expectedByteCount := 21

    if lineCount != expectedLineCount {
        t.Errorf("Expected line count %d, but got %d", expectedLineCount, lineCount)
    } 
    
    if wordCount != expectedWordCount {
        t.Errorf("Expected word count %d, but got %d", expectedWordCount, wordCount)
    } 

    if byteCount != expectedByteCount {
        t.Errorf("Expected byte count %d, but got %d", expectedByteCount, byteCount)
    } 
} 

func TestCountChars(t *testing.T) {
    file, err := os.Create("test.txt")
    if err != nil {
        t.Fatalf("Failed to create the temporary file: %v", err)
    } 
    defer file.Close()
    defer os.Remove("test.txt")

    content := "Sample content with 🚀 emoji"
    _, err = file.WriteString(content)
    if err != nil {
        t.Fatalf("Failed to write to temporary file: %v", err)
    }

    charCount, err := countChars("test.txt")
    if err != nil {
        t.Fatalf("Unexpected error: %v", err)
    }

    byteCount, err := countBytes("test.txt")
    if err != nil {
        t.Fatalf("Unexpected error: %v", err)
    }

    expectedCharCount := 27
    if charCount != expectedCharCount {
        t.Errorf("Expected character count %d, but got %d", expectedCharCount, charCount)
    }

    if charCount == byteCount {
        t.Error("Character count should not be equal to byte count")
    }
} 

func TestCountWords(t *testing.T) {
    file, err := os.Create("test.txt")
    if err != nil {
        t.Fatalf("Failed to create temporary file: %v", err)
    } 
    defer file.Close()
    defer os.Remove("test.txt")

    content := "This is a sample file\nwith multiple words\non each line\n"
    _, err = file.WriteString(content)
    if err != nil {
        t.Fatalf("Failed to write to temporary file: %v", err)
    } 

    count, err := countWords("test.txt")
    if err != nil {
        t.Fatalf("Unexpected error: %v", err)
    } 

    expectedCount := 11
    if count != expectedCount {
        t.Errorf("Expected word count %d, but got %d", expectedCount, count)
    } 
} 

func TestCountLines(t *testing.T) {
    file, err := os.Create("test.txt")
    if err != nil {
        t.Fatalf("Failed to create temporary file: %v", err)
    } 
    defer file.Close()
    defer os.Remove("test.txt")

    content := "Line 1\nLine 2\nLine 3\n"
    _, err = file.WriteString(content)
    if err != nil {
        t.Fatalf("Failed to write to temporary file: %v", err)
    } 

    count, err := countLines("test.txt")
    if err != nil {
        t.Fatalf("Unexpected error: %v", err)
    } 

    expectedCount := 3
    if count != expectedCount {
        t.Errorf("Expected line count %d, but got %d", expectedCount, count)
    } 

}

func TestCountLinesFileNotFound(t *testing.T) {
    _, err := countLines("nonexistent.txt")

    if err == nil {
        t.Error("Expected an error, but got nil")
    } 
} 

func TestCountBytes(t *testing.T) {
    // Create a temp file with sample input
    file, err := os.Create("test.txt")
    if err != nil {
        t.Fatalf("Failed to create temporary file: %v", err)
    }
    defer file.Close()
    defer os.Remove("test.txt")

    content := []byte("Sample content")
    _, err = file.Write(content)
    if err != nil {
        t.Fatalf("Failed to write to temproary file: %v", err)
    } 

    // Call the countBytes function
    count, err := countBytes("test.txt")
    if err != nil {
        t.Fatalf("Unexpted error: %v", err)
    }

    // Assert the expected byte count
    if count != len(content) {
        t.Errorf("Expected byte count %d, but got %d", len(content), count)
    } 
} 

func TestCountBytesFileNotFound(t *testing.T) {
    // Call the countBytes function with a non-existent file
    _, err := countBytes("nonexistent.txt")

    // Assert that the expected error is returned
    if err == nil {
        t.Error("Expected an error, but got nil")
    } 
} 

func TestReadFromStdin(t *testing.T) {
    // Prepare the input data
    input := "Line 1\nLine 2\nLine 3\n"

    // Create a temporary file to store the input data
    tempFile, err := os.CreateTemp("", "test_input")
    if err != nil {
        t.Fatalf("Failed to create temporary file: %v", err)
    }
    defer os.Remove(tempFile.Name())

    // Write the input data to the temporary file
    _, err = tempFile.WriteString(input)
    if err != nil {
        t.Fatalf("Failed to write input data to temporary file: %v", err)
    }
    tempFile.Close()

    // Redirect standard input to the temporary file
    oldStdin := os.Stdin
    defer func() { os.Stdin = oldStdin }()
    file, err := os.Open(tempFile.Name())
    if err != nil {
        t.Fatalf("Failed to open temporary file: %v", err)
    }
    defer file.Close()
    os.Stdin = file

    // capture the std output
    oldStdout := os.Stdout
    defer func() { os.Stdout = oldStdout }()
    r, w, _ := os.Pipe()
    os.Stdout = w

    main()

    w.Close()

    // Read captured output 
    var buf bytes.Buffer 
    _, err = io.Copy(&buf, r)
    if err != nil {
        t.Fatalf("Failed to read captured output: %v", err)
    } 

    // Check the output
    expectedOutput := "3 6 18\n"
    if buf.String() != expectedOutput {
        t.Errorf("Expected output: %q, but got: %q", expectedOutput, buf.String())
    }
}

File: wc-clone/wc-c/wc_test.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

extern int count_bytes(FILE* file);
extern int count_lines(FILE* file);
extern int count_words(FILE* file);
extern int count_chars(FILE* file);

void test_default_count() {
    FILE* file = fopen("test.txt", "w");
    fprintf(file, "Line 1\nLine 2\nLine 3\n");
    fclose(file);

    file = fopen("test.txt", "r");
    int line_count = count_lines(file);
    int word_count = count_words(file);
    int byte_count = count_bytes(file);
    int char_count = count_chars(file);
    fclose(file);

    assert(line_count == 3);
    assert(word_count == 6);
    assert(byte_count == 21);
    assert(char_count == 21);

    remove("test.txt");
} 

void test_count_chars() {
    FILE* file = fopen("test.txt", "w");
    fprintf(file, "Sample content with 🚀 emoji");
    fclose(file);

    file = fopen("test.txt", "r");
    int char_count = count_chars(file);
    int byte_count = count_bytes(file);
    fclose(file);

    assert(char_count == 27);
    assert(byte_count == 30);
    assert(char_count != byte_count);

    remove("test.txt");
} 

void test_count_bytes() {
    FILE* file = fopen("test.txt", "w");
    fprintf(file, "Sample content");
    fclose(file);

    // Call the count_bytes function
    int count = count_bytes(file);

    // Assert the expected byte count
    assert(count = 14);

    // clean up the temp file
    remove("test.txt");
}

void test_count_lines() {
    FILE* file = fopen("test.txt", "w");
    fprintf(file, "Line 1\nLine 2\nLine 3\n");
    fclose(file);

    file = fopen("test.txt", "r");
    int count = count_lines(file);
    fclose(file);

    assert(count == 3);

    remove("test.txt");
}

void test_count_words() {
    FILE* file = fopen("test.txt", "w");
    fprintf(file, "This is a sample file\nwith multiple words\non each line\n");
    fclose(file);

    file = fopen("test.txt", "r");
    int count = count_words(file);
    fclose(file);

    assert(count == 11);

    remove("test.txt");
}

int main() {
    test_count_bytes();
    test_count_lines();
    test_count_words();
    test_count_chars();
    test_default_count();

    printf("All tests passed!\n");

    return 0;
} 
